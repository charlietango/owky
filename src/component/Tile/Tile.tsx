import React, { useEffect, useState } from 'react';
import styled from 'styled-components';
import { LinearGradient } from 'expo-linear-gradient';
import { Text, View } from 'react-native';
import * as Clipboard from 'expo-clipboard';
import * as Haptics from 'expo-haptics';
import { NotificationFeedbackType } from 'expo-haptics';
import {
  Directions,
  FlingGestureHandler,
  FlingGestureHandlerEventPayload,
  HandlerStateChangeEvent,
  LongPressGestureHandler,
  State,
} from 'react-native-gesture-handler';
import chroma from 'chroma-js';

import { generateTokenFromSecret } from '@helper/authenticator';
import ProgressBar from '@component/ProgressBar/ProgressBar';
import IconButton from '@component/IconButton/IconButton';
import { Icon, IconSize } from '@component/IconButton/Icons';
import { Account } from '@type/account';

const Wrapper = styled(View)`
  width: 100%;
  padding: 10px 17px 0 17px;
`;

const Gradient = styled(LinearGradient)`
  border-radius: 17px;
  padding: 15px;
  width: 100%;
  height: 120px;
`;

const ContentWrapper = styled(View)`
  flex-direction: row;
`;

const DetailsWrapper = styled(View)`
  width: 90%;
`;

const ActionsWrapper = styled(View)`
  justify-content: center;
`;

const TokenLabel = styled(Text)`
  font-size: 40px;
  line-height: 48px;
  font-weight: 900;
  color: white;
`;

const UsernameLabel = styled(Text)`
  color: white;
  font-weight: 500;
`;

const IssuerLabel = styled(Text)`
  color: white;
  font-weight: 700;
`;

interface TileProps {
  account: Account;
}

export default function Tile({ account }: TileProps): JSX.Element {
  const [token, setToken] = useState(generateTokenFromSecret(account.secret));
  const [hidden, setHidden] = useState(true);

  useEffect(() => {
    const intervalId = setInterval(() => {
      setToken(generateTokenFromSecret(account.secret));
    }, 1000);

    return () => {
      clearInterval(intervalId);
    };
  }, []);

  const handleFling = (event: HandlerStateChangeEvent<FlingGestureHandlerEventPayload>) => {
    if (event.nativeEvent.state === State.ACTIVE) {
      setHidden(!hidden);
    }
  };

  const handleLongPress = (event: HandlerStateChangeEvent<LongPressGestureHandler>) => {
    if (event.nativeEvent.state === State.ACTIVE) {
      console.log('long press');
    }
  };

  const handleCopy = () => {
    Clipboard.setString(token.token);
    Haptics.notificationAsync(NotificationFeedbackType.Success);
  };

  return (
    <Wrapper>
      <Gradient colors={[chroma(account.color).saturate().hex(), account.color]}>
        <FlingGestureHandler
          direction={Directions.RIGHT | Directions.LEFT}
          onHandlerStateChange={handleFling}
        >
          <LongPressGestureHandler minDurationMs={800} onHandlerStateChange={handleLongPress}>
            <ContentWrapper>
              <DetailsWrapper>
                <TokenLabel>{hidden ? '******' : token.token}</TokenLabel>
                <UsernameLabel>{account.username}</UsernameLabel>
                <IssuerLabel>{account.issuer}</IssuerLabel>
              </DetailsWrapper>
              <ActionsWrapper>
                <IconButton icon={Icon.copy} size={IconSize.large} onPress={handleCopy} />
              </ActionsWrapper>
            </ContentWrapper>
          </LongPressGestureHandler>
        </FlingGestureHandler>
        <ProgressBar timeRemaining={token.timeRemaining} />
      </Gradient>
    </Wrapper>
  );
}
