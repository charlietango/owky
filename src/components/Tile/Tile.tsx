import React from 'react';
import styled from 'styled-components';
import { sample } from 'lodash';
import { LinearGradient } from 'expo-linear-gradient';
import { Text, View } from 'react-native';

const Container = styled(View)`
  width: 100%;
  padding: 10px 10px 0 10px;
`;

const Shape = styled(LinearGradient)<{ color: string }>`
  border-radius: 17px;
  padding: 15px;
  width: 100%;
  height: 120px;
`;

const COLORS = ['#F75D94', '#7929FE', '#FF5E06', '#08D2C5', '#E51B09', '#0D0D0D'];

export default function Tile(): JSX.Element {
  const color = sample(COLORS) as string;

  return (
    <Container>
      <Shape colors={[color + 'AA', color]} color={color}>
        <Text>Pepe</Text>
      </Shape>
    </Container>
  );
}
