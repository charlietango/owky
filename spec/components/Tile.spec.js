import React from 'react';
import renderer from 'react-test-renderer';

const { describe, it, expect } = global;

import Tile from '@components/Tile/Tile';

describe('Test @component <Tile />', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Tile title={'title'} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
