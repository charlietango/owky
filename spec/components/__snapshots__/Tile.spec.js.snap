// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test @component <Tile /> renders correctly 1`] = `
<View
  className="sc-bdfBQB cbLhkM"
>
  <ViewManagerAdapter_ExpoLinearGradient
    proxiedProperties={
      Object {
        "className": "sc-gsTEea hVDAoW",
        "color": "#0D0D0D",
        "colors": Array [
          2852982029,
          4279045389,
        ],
        "endPoint": undefined,
        "locations": undefined,
        "startPoint": undefined,
      }
    }
  >
    <Text>
      title
    </Text>
  </ViewManagerAdapter_ExpoLinearGradient>
</View>
`;
